packages/@n8n/nodes-langchain/nodes/tools/ToolHttpRequest/utils.ts

```diff
@@ -0,0 +1,793 @@
+
+	if (inputType === 'json') {
+		rawRequestOptions[requestOptionsProperty] = ctx.getNodeParameter(
+			jsonPropertyName,
+			itemIndex,
+			'',
+		) as string;
+	} else {
+		parametersValues = ctx.getNodeParameter(
+			parametersPropertyName,
+			itemIndex,
+			[],
+		) as ParametersValues;
+	}
+
+	const inputParameters = prepareParameters(
+		parametersValues,
+		placeholdersDefinitions,
+		inputType,
+		requestOptionsProperty,
+		MODEL_INPUT_DESCRIPTION[requestOptionsProperty],
+		rawRequestOptions[requestOptionsProperty],
+	);
+
+	toolParameters.push(...inputParameters.parameters);
+
+	requestOptions[requestOptionsProperty] = {
+		...(requestOptions[requestOptionsProperty] as IDataObject),
+		...inputParameters.values,
+	};
+};
+
+export const prepareToolDescription = (
+	toolDescription: string,
+	toolParameters: ToolParameter[],
+) => {
+	let description = `${toolDescription}`;
+
+	if (toolParameters.length) {
+		description += `
+	Tool expects valid stringified JSON object with ${toolParameters.length} properties.
+	Property names with description, type and required status:
+
+	${toolParameters
+		.filter((p) => p.name)
+		.map(
+			(p) =>
+				`${p.name}: (description: ${p.description || ''}, type: ${p.type || 'string'}, required: ${!!p.required})`,
+		)
+		.join(',\n ')}
+
+		ALL parameters marked as required must be provided`;
+	}
+
+	return description;
+};
+
+export const configureToolFunction = (
+	ctx: IExecuteFunctions,
+	itemIndex: number,
+	toolParameters: ToolParameter[],
+	requestOptions: IHttpRequestOptions,
+	rawRequestOptions: { [key: string]: string },
+	httpRequest: (options: IHttpRequestOptions) => Promise<any>,
+	optimizeResponse: (response: string) => string,
+) => {
+	return async (query: string): Promise<string> => {
+		const { index } = ctx.addInputData(NodeConnectionType.AiTool, [[{ json: { query } }]]);
+
+		let response: string = '';
+		let options: IHttpRequestOptions | null = null;
+		let executionError: Error | undefined = undefined;
+
+		if (!toolParameters.length) {
+			query = '{}';
+		}
+
+		try {
+			if (query) {
+				let dataFromModel;
+
+				try {
+					dataFromModel = jsonParse<IDataObject>(query);
+				} catch (error) {
+					if (toolParameters.length === 1) {
+						dataFromModel = { [toolParameters[0].name]: query };
+					} else {
+						throw new NodeOperationError(
+							ctx.getNode(),
+							`Input is not a valid JSON: ${error.message}`,
+							{ itemIndex },
+						);
+					}
+				}
+
+				for (const parameter of toolParameters) {
+					if (
+						parameter.required &&
+						(dataFromModel[parameter.name] === undefined || dataFromModel[parameter.name] === null)
+					) {
+						throw new NodeOperationError(
+							ctx.getNode(),
+							`Model did not provide parameter '${parameter.name}' which is required and must be present in the input`,
+							{ itemIndex },
+						);
+					}
+				}
+
+				options = requestOptions;
+
+				for (const parameter of toolParameters) {
```

Comment:
Do we need to have two separate loops here?

Reply:
yes, better to check if model provided all required parameters before actually doing any processing